Fix some edge cases for stringifyBody in rsa signature generation.

lime008 opened the pull request:
This pull request fixes the following issue and adds a small optimization to the shortObject function. Unnecessary lodash dependency was also removed.

When preparing the JSON formatting for the signature calculation a regular expression is used to replace the character sequence `":` with `": ` which will break if the string value contains an escaped quote followed with a colon character.

Example:
```json
// input
{
    "foo":"test value\":hello"
}

// expected output
{"foo": "test value\":hello"}

// what happens
{"foo": "test value\": hello"}
//                    ^
// notice the added space character here
```

Commit:
optimize sortObject

Commit:
fix formatting

Commit:
more robust way to adding spaces after json key-value separators

Commit:
fix formatting

Commit:
more formatting fixes

Commit:
remove lodash dependency

hoanvancs commented on app/lib/rsa.js:
```diff
@@ -84,24 +94,22 @@ const sendPublicKey = function (req, res) {
+				sortedObj[key] = object[key]
+				return
+			}
+			sortedObj[key] = sortObjectAlt(object[key])
```
@lime008 Where is `sortObjectAlt` function? I do not see it anywhere in project.

lime008 replied:
Oh must have forgot to change the function call when I replaced the original with the new one after testing. I'll also add the tests to the pull request.

hoanvancs commented on app/lib/rsa.js:
```diff
@@ -123,13 +123,43 @@ const sortObject = function(object) {
-        )
-        .replace(new RegExp('":', "g"), '": ')
-        .trim();
+    const json = JSON.stringify(sortObject(body));
```
@lime008 Can you create a unit-test for the new changed function? We can ensure both returns the same output.

Commit:
Add unit tests for the stringifyBody

Also fixed a wrong character code value found in testing.

Old implementations of stringifyBody and sortObject are now included in the rsa.test.js but can be
removed when the new ones have been validated. There is a way to make the old
implementation failures to pass the test suite, but they are disabled
for now to show the pitfalls of the old implementation.

lime008 commented:
@hoanvancs I added the unit-tests to the latest commit.

hoanvancs commented on test/rsa.test.js:
```diff
@@ -1,90 +1,209 @@
+            expect: '{"test": "\\u00e4\\u00f6uy\\u00e5\\u00c5"}',
+        },
+        {
+            name: 'should succeed with no added space and fail with the old implementation',
```
I am trying to run the tests and this test case fails.

hoanvancs commented on test/rsa.test.js:
```diff
@@ -1,90 +1,209 @@
+            shouldFailOnOldImplementation: true,
+        },
+        {
+            name: 'should succeed with non object value and fail with the old implementation',
```
And this test case also fail.

lime008 commented:
@hoanvancs This behaviour is expected and is detailed in the full commit message: 
> Old implementations of stringifyBody and sortObject are now included in the rsa.test.js but can be
> removed when the new ones have been validated. There is a way to make the old
> implementation failures to pass the test suite, but they are disabled
> for now to show the pitfalls of the old implementation.

lime008 commented on test/rsa.test.js:
```diff
@@ -1,90 +1,209 @@
+                out.to.throw();
+                return;
+            }
+            /** Uncomment this to make the test suite pass with the old implementation
```
@hoanvancs You can uncomment this comment block to make the tests pass with the old implementation because these are the cases that the old implementation doesn't support.

hoanvancs commented:
@artoranta Could you please take a look into this PR?

artoranta commented:
> @artoranta Could you please take a look into this PR?

Looks really good. I'm glad to see this issue was discovered, great work. Merge is ok.

hoanvancs commented:
> > @artoranta Could you please take a look into this PR?
> 
> Looks really good. I'm glad to see this issue was discovered, great work. Merge is ok.

Thank @artoranta
